### ISSUE TITLE
[aggregatestore/events] aggregate version handling

### ISSUE THREAD

**Issue by grillbiff** [looplab/eventhorizon#17]
I can't help to think the version handling in the event store is wrong. (using the mongodb implementation as example)
If I understand the code correctly it only makes sure that the aggregate has not changed from when it is read

```
    for _, event := range events {
        // Get an existing aggregate, if any.
        var existing []mongoAggregateRecord
        err := sess.DB(s.db).C("events").FindId(event.AggregateID().String()).
            Select(bson.M{"version": 1}).Limit(1).All(&existing)
```

until it is written

```
            // Increment record version before inserting.
            r.Version = existing[0].Version + 1

            // Increment aggregate version on insert of new event record, and
            // only insert if version of aggregate is matching (ie not changed
            // since the query above).
            err = sess.DB(s.db).C("events").Update(
                bson.M{
                    "_id":     event.AggregateID().String(),
                    "version": existing[0].Version,
                },
                bson.M{
                    "$push": bson.M{"events": r},
                    "$inc":  bson.M{"version": 1},
                },
```

The purpose of the version check should be to make sure that the event is applied (saved) to the correct version of the aggregate when the event was first created. That is, the aggregate version that should be validated is the one that was instantiated when the command is applied and the event(s) is created. 

A solution would be to add a Version attribute to the `eventhorizon.Event` which holds the aggregate version which the event should be applied to. This can then be validated when storing the event.

Hope it makes sense, can be a bit hairy to explain :)

**Comment by maxekman** [comment]
I think I understand you correctly, and the implementation may very well have issues with this (as it's still not considered production hardened).

An event that is to be saved should only be allowed to do so if the version of the aggregate has not been changed by another event write in that could have happened in-between, in which case the event version would be incorrect. In this case we simply refuse writing, which is may or may not be the desired solution long term. The version field on the event you are suggesting is already existing per implementation and used for knowing which version of the aggregate the event represents (there can only be one event per version). Enforcing a `Version()` method on the interface could maybe be helpful to do better version checking in other places.

A good solution for MongoDB could be to increment the aggregate version and use that version in the event in the same atomic operation (including using 1 if the aggregate is new). Not sure how to do that right now.

Any thoughts on this? Did I understand you correctly?

**Comment by maxekman** [comment]
I would love to harden the versioning btw, thanks for bringing up the discussion!

**Comment by grillbiff** [comment]
Yes I think you understood me but I´m still quite new to CRQRS/ES so I might be totally off here :)

```
- In the aggregate
1. An aggregate with version 2 is loaded
2. A command is handled by the aggregates command handler which generates an event with version 2 stating that it can only be applied to the aggregate if it is in version 2
3. the generated event is saved to the event store
- In the event store 
4. The aggregate for the event which is beeing saved is loaded
5. The aggregate version is checked against the event version. If they match the event is committed and the aggregate version is bumped. If the aggregate version does not match for some reason (another event from another process was committed) an error is returned. The usual way of handling this (as far as my understanding goes) is to simply rerun the command which generated the denied event but this will ultimately be up to the biz rules.
```

**Comment by maxekman** [comment]
It is slightly different, the events version is the version of the aggregate _after_ the event is applied. Also the aggregate version is only checked against the saved aggregate version, not the event. This only prevents the event to be saved if another event is save that has "taken" the version that is being saved. In this case the strategy would be to try to save again, in which case the version would be one higher due to the other event that "took" our version. Makes sense?

In any case, an events version is what version the aggregate is at after the event is applied, not which version to apply it to.

**Comment by grillbiff** [comment]
yes, so I think a version checking of the aggregate version at the time of when the biz rules are evaluated (aka command is handled) when the event is saved. This would also cover the error scenario you described above. 

If a version check fails, the command has to be evaluated again againts the updated aggregate, if we simply bump the event version and save it it it is never validated against the new state of the aggregate.

So in a bank account scenario I could end up with a negative balance even though the business rules prohibits it if I simply reuse the events. What I should do is rerun the command.

**Comment by maxekman** [comment]
Makes sense, let me think about this. If you have time you are welcome to make a PR with a proposed solution. Make sure to update the examples if you do.

**Comment by grillbiff** [comment]
Yes, we are still deciding on a CQRS/ES framework for our next project but if we go ahead with eventhorizon I will definitely be able to contribute.

**Comment by maxekman** [comment]
That is awesome! Hope you decide on Event Horizon. I'm soon coming back to work on the project where it was initially conceived, that will definitely mean a lot of changes from me again (have been doing other projects for a while).

**Comment by maxekman** [comment]
I did some work on this in #19, please take a look and leave some comments on the approach if you want. This article contained a few useful bits: https://lostechies.com/gabrielschenker/2015/07/13/event-sourcing-applied-the-repository/

**Comment by grillbiff** [comment]
Yes, read that as well, I'll have a look at the PR

**Comment by maxekman** [comment]
Some more progress is done on this in #65 and #66.

**Comment by maxekman** [comment]
Fixed in #63.
